hsds/domain_crawl.py

```diff
@@ -0,0 +1,275 @@
+                msg = f"POST attributes requested {len(attr_names)}, "
+                msg += f"but only {len(attributes)} were returned"
+                log.warn(msg)
+            self._obj_dict[obj_id] = attributes
+        else:
+            log.warn(f"Domain crawler - got {status} status for obj_id {obj_id}")
+            self._obj_dict[obj_id] = {"status": status}
+       
+
+    async def put_attributes(self, obj_id, attr_items):
+        # get the given attributes for the obj_id
+        log.debug(f"put_attributes for {obj_id}, {len(attr_items)} attributes")
+        req = getDataNodeUrl(self._app, obj_id)
+        collection = getCollectionForId(obj_id)
+        req += f"/{collection}/{obj_id}/attributes"
+        params = {}
+        if "bucket" in self._params:
+            params["bucket"] = self._params["bucket"]
+        data = {"attributes": attr_items}
+        status = None
+        put_rsp = None
+        try:
+            put_rsp = await http_put(self._app, req, data=data, params=params)
+        except HTTPConflict:
+            log.warn("DomainCrawler - got HTTPConflict from http_put")
+            status = 409
+        except HTTPServiceUnavailable:
+            status = 503
+        except HTTPInternalServerError:
+            status = 500
+        except Exception as e:
+            log.error(f"unexpected exception {e}")
+
+        if put_rsp is not None:
+            log.info(f"PUT Attributes resp: {put_rsp}") 
+            if "status" in put_rsp:
+                status = put_rsp["status"]
+            else:
+                status = 201
+        log.debug(f"DomainCrawler fetch for {obj_id} - returning status: {status}")
+        self._obj_dict[obj_id] = {"status": status}
+         
+    async def get_obj_json(self, obj_id):
+        """ get the given obj_json for the obj_id.  
+            for each group found, search the links if include_links is set """
+        kwargs = {}
+        for k in ("include_links", "include_attrs", "bucket"):
+            if k in self._params:
+                kwargs[k] = self._params[k]
+         
+        obj_json = None
+        status = 200
+        try:
+            obj_json = await getObjectJson(self._app, obj_id, **kwargs)
+        except HTTPNotFound:
+            status = 404
+        except HTTPServiceUnavailable:
+            status = 503
+        except HTTPInternalServerError:
+            status = 500
+        except Exception as e:
+            log.error(f"unexpected exception {e}")
+            status = 500
+    
+        if obj_json is None:
+            msg = f"DomainCrawler - getObjectJson for {obj_id} "
+            if status >= 500:
+                msg += f"failed, status: {status}"
+                log.error(msg)
+            else:
+                msg += f"returned status: {status}"
+                log.warn(msg)
+            return
+        
+        log.debug(f"DomainCrawler - got json for {obj_id}")
+
+        # if including links, we need link count
```

Comment:
Should be `we don't need link count`?

Reply:
Thinking was you don't need it since you can jsut do len(links), but added back to have fewer special cases.